Allow skipping redirects for extensions.

dereuromark opened the pull request:
Resolves https://github.com/cakephp/authentication/issues/717

This happens when you have default (admin) routing and just add extensions into it in general.
You cannot then control for what actions only.
And if you then get redirects on those, that is weird (API etc) and creates overhead on the server.

dereuromark committed:
Allow skipping redirects for extensions.

dereuromark committed:
Allow skipping redirects for extensions.

dereuromark committed:
Allow skipping redirects for extensions.

markstory commented on src/Middleware/UnauthorizedHandler/RedirectHandler.php:
```diff
@@ -44,6 +44,7 @@ class RedirectHandler implements HandlerInterface
         'url' => '/login',
         'queryParam' => 'redirect',
         'statusCode' => 302,
+        'allowedRedirectExtensions' => false,
```
We should add this new option to the docblock.

dereuromark replied:
Done
I switched the default to true, as that is more logical.

markstory commented on docs/en/middleware.rst:
```diff
@@ -198,6 +198,10 @@ Both redirect handlers share the same configuration options:
 * ``queryParam`` - the accessed request URL will be attached to the redirect URL
   query parameter (``redirect`` by default).
 * ``statusCode`` - HTTP status code of a redirect, ``302`` by default.
+* ``allowedRedirectExtensions`` - an array of allowed file extensions for redirecting.
+  If the request URL has a file extension that is not in this list, the redirect will not
+  happen and the exception will be rethrown. This is useful to prevent unauthorized access
+  to API based responses, that should not be redirecting in any case. `false` by default and not enabled then.
```
```suggestion
* ``allowedRedirectExtensions`` - an array of allowed file extensions for redirecting.
  If the request URL has a file extension that is not in this list, the redirect will not
  happen and the exception will be rethrown. Can also be a boolean to toggle on/off
  redirects entirely. This is useful to prevent unauthorized access to API based 
  responses, that should not be redirecting in any case. `false` by default and not enabled then.
```

markstory commented on tests/TestCase/Middleware/UnauthorizedHandler/RedirectHandlerTest.php:
```diff
@@ -160,4 +160,48 @@ public function testHandleException()
+        ]);
+    }
+
+    public function testHandleRedirectionWithExtensionWhitelisted(): void
```
```suggestion
    public function testHandleRedirectionWithExtensionAllowlisted(): void
```

markstory commented on tests/TestCase/Middleware/UnauthorizedHandler/RedirectHandlerTest.php:
```diff
@@ -160,4 +160,48 @@ public function testHandleException()
+    {
+        $handler = new RedirectHandler();
+
+        $exception = new Exception();
+        $request = ServerRequestFactory::fromGlobals(
+            ['REQUEST_METHOD' => 'GET'],
+        );
+        $request = $request->withParam('_ext', 'csv');
+
+        $this->expectException(Exception::class);
+
+        $handler->handle($exception, $request, [
+            'exceptions' => [
+                Exception::class,
```
```suggestion
        $exception = new LogicException();
        $request = ServerRequestFactory::fromGlobals(
            ['REQUEST_METHOD' => 'GET'],
        );
        $request = $request->withParam('_ext', 'csv');

        $this->expectException(LogicException::class);

        $handler->handle($exception, $request, [
            'exceptions' => [
                LogicException::class,
```

Any concern on this being a broad type expectation? You could use something more narrow `LogicException` and have more tightly scoped behavior captured in the test.

dereuromark committed:
Update docs/en/middleware.rst

Co-authored-by: Mark Story <mark@mark-story.com>

dereuromark committed:
Update tests/TestCase/Middleware/UnauthorizedHandler/RedirectHandlerTest.php

Co-authored-by: Mark Story <mark@mark-story.com>

dereuromark committed:
Update tests/TestCase/Middleware/UnauthorizedHandler/RedirectHandlerTest.php

Co-authored-by: Mark Story <mark@mark-story.com>

dereuromark committed:
Allow skipping redirects for extensions.

markstory commented on src/Exception/LogicException.php:
```diff
@@ -0,0 +1,70 @@
+use Authorization\Policy\ResultInterface;
+use Throwable;
+
+class LogicException extends Exception
```
I thought you would have used the stdlib `LogicException` but this works too. :shrug:  Edit: I see now that my suggestion is not possible. I'll push a commit up instead of leaving more comments.

markstory commented on src/Exception/LogicException.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return \Authorization\Policy\ResultInterface|null
+     */
+    public function getResult(): ?ResultInterface
```
I don't think this method will be used.

markstory committed:
Remove LogicException

- We can use one of the existing exceptions from the plugin.
- Expand test coverage to cover all the new branches
